src/community/features-autopopulate/features-autopopulate-core/src/main/java/org/geoserver/autopopulate/AutopopulateTransactionCallback.java

```diff
@@ -0,0 +1,407 @@
+     */
+    @Override
+    public void afterTransaction(
+            final TransactionRequest request, TransactionResponse result, boolean committed) {
+        // nothing to do
+    }
+
+    /**
+     * @return {@code 0}, we don't need any special treatment
+     * @see TransactionCallback#getPriority()
+     */
+    @Override
+    public int getPriority() {
+        return ExtensionPriority.LOWEST;
+    }
+
+    /**
+     * Collects "transactionCustomizer.properties" file that contains a set of names and CQL
+     * expressions, e.g.:
+     *
+     * <pre>
+     *  UTENTE=env('GSUSER')
+     *  AGGIORNAMENTO=now()
+     *  </pre>
+     *
+     * <p>To keep things simple, the expressions will just use environment variables, but not see
+     * the other values provided in the update/insert, and will not be differentiated by
+     * insert/update cases.
+     *
+     * @see TransactionCallback#beforeTransaction(TransactionRequest)
+     */
+    @Override
+    public TransactionRequest beforeTransaction(TransactionRequest request) throws WFSException {
+        if (request.getVersion() == null) {
+            return request;
+        }
+
+        List<TransactionElement> newElements = new ArrayList<>();
+        for (TransactionElement element : request.getElements()) {
+            if (element instanceof Insert) {
+                List<SimpleFeature> newFeatures = new ArrayList<>();
+                for (Object of : affectedFeatures(element)) {
+                    if (of instanceof SimpleFeature) {
+                        try {
+                            LOGGER.info("Updating feature: " + of);
+                            SimpleFeature transformed = applyTemplate((SimpleFeature) of);
+                            newFeatures.add(transformed);
+                        } catch (RuntimeException | IOException e) {
+                            // Do never make the transaction fail due to an
+                            // AutopopulateTransactionCallback error.
+                            // Yell on the logs though
+                            LOGGER.log(
+                                    Level.WARNING,
+                                    "Error pre computing the transaction's affected attributes",
+                                    e);
+                            newFeatures.add((SimpleFeature) of);
+                        }
+                    }
+                }
+                Insert insertElement = (Insert) element;
+                insertElement.setFeatures(newFeatures);
+                newElements.add(element);
+            } else if (element instanceof Update) {
+                FeatureTypeInfo featureTypeInfo =
+                        getFeatureTypeInfo(new NameImpl(element.getTypeName()));
+                try {
+                    Update updateElement = (Update) element;
+                    SimpleFeature feature =
+                            getTransactionSource(updateElement).getFeatures().features().next();
```

Comment:
This will leak a feature iterator, eventually starving connection pools. Use ``DataUtilities.first(getTransactionSource(updateElement).getFeatures())`` instead.

Reply:
Yessir, ty. Done!